### DISCUSSION TITLE
Better IDE type hint for Depends

### DISCUSSION THREAD

**Discussion by 0x587** [fastapi/fastapi#12113] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [x] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
T = TypeVar("T")
def Depends(  # noqa: N802
        dependency: Annotated[
            Optional[
                Union[Callable[..., Coroutine[Any, Any, T]], Callable[..., T]]
            ],
            Doc(
                """
                A "dependable" callable (like a function).
    
                Don't call it directly, FastAPI will call it for you, just pass the object
                directly.
                """
            ),
        ] = None,
        *,
        use_cache: Annotated[
            bool,
            Doc(
                """
                By default, after a dependency is called the first time in a request, if
                the dependency is declared again for the rest of the request (for example
                if the dependency is needed by several dependencies), the value will be
                re-used for the rest of the request.
    
                Set `use_cache` to `False` to disable this behavior and ensure the
                dependency is called again (if declared more than once) in the same request.
                """
            ),
        ] = True,
) -> T:
    from fastapi import Depends
    return Depends(dependency, use_cache=use_cache)

def f1() -> int:
    return 1


async def f2() -> int:
    return 1


v1 = Depends(f1)
v2 = Depends(f2)
```

### Description

I modified the type definition of Depends, perhaps this will provide better IDE type inference.
![image](https://github.com/user-attachments/assets/98a67215-b932-4c24-8b54-7f6071b4a9f2)
![image](https://github.com/user-attachments/assets/c2f97a10-f7c4-44a6-98d3-ab874fb0c45a)

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.112.2

### Pydantic Version

2.8.2

### Python Version

Python 3.12.4

**Comment by YuriiMotov** [comment, marked as answer]
It's recommended now to use the approach with Annotated:
```py
@router.post("/login")
def login(
  database: Annotated[DBSession, Depends(get_database)],
):
  pass
```

Generic `Depends` will not help when `Annotated` approach is used.

There are a couple of PR that aimed to implement this, but thy were closed with that reason:
* https://github.com/fastapi/fastapi/pull/11255
* https://github.com/fastapi/fastapi/pull/10053
